zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/transform/DeploymentTransformer.java

```diff
@@ -111,54 +111,68 @@ public Either<Failure, Void> transform(final DeploymentRecord deploymentEvent) {
 
     // step 1: only validate the resources and add their metadata to the deployment record (no event
     // records are being written yet)
+    final var bpmnResources = new ArrayList<BpmnResource>();
     while (resourceIterator.hasNext()) {
       final DeploymentResource deploymentResource = resourceIterator.next();
-      success &=
-          transformResource(
-              deploymentEvent,
-              errors,
-              deploymentResource,
-              transformer -> transformer::createMetadata);
+      if (isBpmnResource(deploymentResource)) {
+        final var context = new BpmnElementsWithDeploymentBinding();
+        bpmnResources.add(new BpmnResource(deploymentResource, context));
+        success &= createMetadata(deploymentResource, deploymentEvent, context, errors);
+      } else {
+        success &= createMetadata(deploymentResource, deploymentEvent, null, errors);
+      }
+    }
+
+    // intermediate step (for BPMN resources only): validate process elements that use deployment
+    // binding (all linked resources must be part of the current deployment)
+    if (success && !bpmnResources.isEmpty()) {
+      final var validator = new BpmnDeploymentBindingValidator(deploymentEvent);
+      for (final var bpmnResource : bpmnResources) {
+        final var validationError = validator.validate(bpmnResource.elements);
+        if (validationError != null) {
+          success = false;
+          errors
+              .append("\n'")
+              .append(bpmnResource.resource.getResourceName())
+              .append("':\n")
+              .append(validationError);
+        }
+      }
     }
 
     // step 2: update metadata (optionally) and write actual event records
     if (success) {
       for (final DeploymentResource deploymentResource : deploymentEvent.resources()) {
-        success &=
-            transformResource(
-                deploymentEvent,
-                errors,
-                deploymentResource,
-                transformer -> transformer::writeRecords);
+        final var transformer = getResourceTransformer(deploymentResource.getResourceName());
+        transformer.writeRecords(deploymentResource, deploymentEvent);
```

Comment:
❓ It is not entirely clear why we don't need a transformation here, could you please explain this change in more details? Was it already unnecessary in the previous implementation?

Reply:
Good question :). I tried to explain the reasoning behind this change in the description of https://github.com/camunda/camunda/pull/22434/commits/d9c86ffbfe05632d479542c9c0509142606b68da.

The `writeRecords` methods were always returning `Either.right(null)` (so never a `Failure`) already before which means that they could never change `success` to `false` anyway. So it seemed fine to make them just return `void`.

Another problem was that `DeploymentResourceTransformer.createMetadata` now has three method parameters instead of two, so the `BiFunction` here did not work any more for both `createMetadata` and `writeRecords`:

https://github.com/camunda/camunda/blob/59b0b3e0e845b11a32e36c0e33c23b17878ccf2d/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/transform/DeploymentTransformer.java#L148-L155

I think it would not have been easily possible to make the `transformResource` method work for both steps (`createMetadata` and `writeRecords`).

Reply:
What I just realized is that the code does not handle any unexpected errors that might occur in `writeRecords` any more (only for `createMetadata`):

https://github.com/camunda/camunda/blob/ad07c9d152d7ec346f552d9cbdca8ac8444cd1e9/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/transform/DeploymentTransformer.java#L185-L187

I'll fix this and add the same kind of exception handling.

Reply:
@jfriedenstab Thank you for the explanation and for adding error handling 👍